locator/abstract_replication_strategy.cc

```diff
@@ -330,27 +330,79 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
 
```

Comment:
> There shouldn't be too much complexity in this optimization, it's just boilerplate'y.

By 'complexity' I meant exactly this - code duplication with tiny tricky differences. This will make it harder to change it in the future.

> When natural_endpoints_depend_on_token was introduced, it dropped startup time from 3000+ seconds to 40 seconds in some clusters. It was significantly affecting some customers. See https://github.com/scylladb/scylladb/commit/9167b857e9c3f32032b6e9a6eed73610c0da0a85 commit message for references.

1. this code is not used at startup, only during topology changes
2. I don't see what the concrete source of perf gains could be here; in 'unoptimised' version we use this code to create natural endpoints
```
    auto eps = tm.get_all_endpoints();
    return make_ready_future<endpoint_set>(endpoint_set(eps.begin(), eps.end()));
```
in 'optimised' we just copy it from another one, not a big deal it seems.

Reply:
We do save some calculations in my version: calculating endpoints.diff and comparing endpoints.target with endpoints.current happens only once, not repeated for every token.

>  in 'unoptimised' version we use this code to create natural endpoints

well, that argument relies on the particular implementation of everywhere_strategy, which today indeed doesn't perform too many calculations. I'm not sure we should rely on the implementation like that.

We could potentially reduce the space/time complexity of this code in the future, right now it's O(num_tokens * num_replicas) due to all the copies of replica sets, maybe we could share the replica sets between different entries of `replication_map` etc. and reduce the complexity to O(num_tokens + num_replicas).

Reply:
1. This optimisation isn't essential to this PR. It was not implemented before, and the PR is about moving the calculation into erm, not optimising it.
2. I'm still not convinced it worth it. I suspect in https://github.com/scylladb/scylladb/commit/9167b857e9c3f32032b6e9a6eed73610c0da0a85 it worked because of eliminated `std::unordered_map`-s construction, since it's very inefficient. BTW, why we don't use abseil's flat_map by default?

Reply:
I've implemented the optimisation in a different way.